### PR TITLE
feat: updated the prepare_for_training methods

### LINKED ISSUE #1697: `prepare_for_training` method should fail if no training data is available
**Describe the bug**

**To Reproduce**
1. Log some records without annotation info into Rubrix
2. Load and apply a training transformation using `prepare_for_training`.

An empty dataset is created.

**Expected behavior**
An error should be raised when no training data is available since nothing can be done for those cases.

**Screenshots**
Screenshot from 2022-08-24 09-36-45

**Environment (please complete the following information):**
 - OS [e.g. iOS]:
 - Browser [e.g. chrome, safari]:
 - Rubrix Version [e.g. 0.10.0]:
 - ElasticSearch Version [e.g. 7.10.2]:
 - Docker Image (optional) [e.g. rubrix:v0.10.0]:  

**Additional context**

### LINKED ISSUE #2045: feat: directly integrate `rg.load`or `rg.Database`, `prepare_for_training` with train test splits
**Is your feature request related to a problem? Please describe.**
To create a train-test split, I now need to rely on external packages.

**Describe the solution you'd like**
I would like to load a dataset which is automatically split into train-test-val data. And potentially, I would like to store the most recent split info under metadata.split.

**Describe alternatives you've considered**
N.A.

**Additional context**
Especially with the integration with spaCy it is difficult.

### LINKED ISSUE #2122: add `prepare_for_training` for `spacy` to `DatasetForTextClassification`
**Is your feature request related to a problem? Please describe.**
I want to train a `DatasetForTextClassification` using `spacy`-

**Describe the solution you'd like**
I would like to have an option to do `DatasetForTextClassification.prepare_for_training(framework="spacy")`

**Describe alternatives you've considered**
Manually creating a dataset.

**Additional context**
#2107

### LINKED ISSUE #2131: add `prepare_for_training` for `sparknlp`
**Is your feature request related to a problem? Please describe.**
There is no integration with `sparknlp`

**Describe the solution you'd like**
I would like to see a better integration of `sparknlp` with Argilla.
"[David Berenstein](https://www.linkedin.com/in/ACoAABqu7N4Blkso-JPgKAvQ3nfTRXOJqJMhV70) [Daniel Vila Suero](https://www.linkedin.com/in/ACoAAAmaxuQB0yuvejYin78IVGuD5-IXhPUDu-g) hey.. you can probably integrate your solution with Spark NLP pipelines as well.. please see this blogpost to see several deployment solutions supported
https://medium.com/spark-nlp/deploying-spark-nlp-for-healthcare-from-zero-to-hero-88949b0c866d
and here are all the healthcare NLP related notebooks
https://github.com/JohnSnowLabs/spark-nlp-workshop/tree/master/tutorials/Certification_Trainings/Healthcare"

**Describe alternatives you've considered**
N.A.

**Additional context**
https://www.linkedin.com/feed/update/urn:li:activity:7016072187870646272/?commentUrn=urn%3Ali%3Acomment%3A(activity%3A7016072187870646272%2C7016076058416295936)&dashCommentUrn=urn%3Ali%3Afsd_comment%3A(7016076058416295936%2Curn%3Ali%3Aactivity%3A7016072187870646272)&dashReplyUrn=urn%3Ali%3Afsd_comment%3A(7016095045325856768%2Curn%3Ali%3Aactivity%3A7016072187870646272)&replyUrn=urn%3Ali%3Acomment%3A(activity%3A7016072187870646272%2C7016095045325856768)

### LINKED ISSUE #2132: `prepare_for_training` does not work for multi-label dataset
**Describe the bug**
I cannot use multi-label dataset.`prepare_for_training` directly.

**To Reproduce**
1. Go to any multi-label dataset.
2. export the dataset via `prepare_for_training`
3. use for training directly

**Expected behavior**
multi-label datasets ought to be delivered with `https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html` or they cannot be used for training

**Screenshots**
N.A.

**Environment (please complete the following information):**
 - OS [e.g. iOS]: N.A.
 - Browser [e.g. chrome, safari]: N.A.
 - Argilla Version [e.g. 1.0.0]:1.1.1
 - ElasticSearch Version [e.g. 7.10.2]:N.A.
 - Docker Image (optional) [e.g. argilla:v1.0.0]:N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

I updated the prepare for training methods.

required for #2154 
Closes #2132 
Closes #2122 
Closes #2045
Closes #1697 

**Type of change**

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [] Documentation update

**How Has This Been Tested**

Please describe the tests that you ran to verify your changes. And ideally reference `tests`.

- [X] prepare_for_training
- [X] prepare_for_training_with_spacy

### CODE REVIEW

**Comment by davidberenstein1957** [comment]
@frascuchon feel free to do an initial review. I still need to update the documentation and I was considering adding #2131 (depending on the effort it will take)

**Comment by davidberenstein1957** [comment]
I will also include this #1697

File: src/argilla/client/datasets.py [hunk 1]
```diff
@@ -360,16 +368,172 @@ def _from_pandas(cls, dataframe: pd.DataFrame) -> "Dataset":
...
+                train_size=train_size, test_size=test_size
+            )
+        elif framework is Framework.SPACY and lang is None:
+            raise ValueError(
+                "Please provide a spacy language model to prepare the"
+                " dataset for training with the spacy framework."
+            )
+        elif framework is Framework.SPACY or framework is Framework.SPARK_NLP:
+            if train_size or test_size:
+                from sklearn.model_selection import train_test_split
```
Comment (frascuchon) [inline src/argilla/client/datasets.py:488]: do we have this defined as a dep?

File: src/argilla/client/datasets.py [hunk 2]
```diff
@@ -360,16 +368,172 @@ def _from_pandas(cls, dataframe: pd.DataFrame) -> "Dataset":
...
+                    test_df = self._prepare_for_training_with_spark_nlp(records_test)
+
+                    return train_df, test_df
+            else:
+                if framework is Framework.SPACY:
+                    return self._prepare_for_training_with_spacy(
+                        nlp=lang, records=self._records
+                    )
+                else:
+                    return self._prepare_for_training_with_spark_nlp(
```
Comment (frascuchon) [inline src/argilla/client/datasets.py:516]: this method should be defined as abstract method in the base class

File: src/argilla/client/datasets.py [hunk 3]
```diff
@@ -360,16 +368,172 @@ def _from_pandas(cls, dataframe: pd.DataFrame) -> "Dataset":
...
+
+                records_train, records_test = train_test_split(
+                    self._records, train_size=train_size, test_size=test_size
+                )
+                if framework is Framework.SPACY:
+                    if lang is None:
+                        raise ValueError(
+                            "Please provide a spacy language model to prepare the"
+                            " dataset for training with the spacy framework."
+                        )
```
Comment (frascuchon) [inline src/argilla/client/datasets.py]: I think this code is duplicated. Could we remove it?

File: src/argilla/client/datasets.py [hunk 4]
```diff
@@ -360,16 +368,172 @@ def _from_pandas(cls, dataframe: pd.DataFrame) -> "Dataset":
...
+        if framework is Framework.TRANSFORMERS:
+            return self._prepare_for_training_with_transformers(
+                train_size=train_size, test_size=test_size
+            )
+        elif framework is Framework.SPACY and lang is None:
+            raise ValueError(
+                "Please provide a spacy language model to prepare the"
+                " dataset for training with the spacy framework."
+            )
+        elif framework is Framework.SPACY or framework is Framework.SPARK_NLP:
```
Comment (frascuchon) [inline src/argilla/client/datasets.py]:
```suggestion
        elif framework in [Framework.SPACY, Framework.SPARK_NLP]:
```

File: src/argilla/client/datasets.py [hunk 5]
```diff
@@ -360,16 +368,172 @@ def _from_pandas(cls, dataframe: pd.DataFrame) -> "Dataset":
...
+            if train_size or test_size:
```
Comment (frascuchon) [inline src/argilla/client/datasets.py:487]: Could we assume a default value of `1.0` for `train_size` and always use the `train_test_split` function? By default, this function applies a `shuffle` that may be cool to apply for preparing the training data. What do you think?

File: tests/client/test_dataset.py [hunk 6]
```diff
@@ -600,6 +615,45 @@ def test_prepare_for_training_with_spacy(self):
         assert isinstance(train, spacy.tokens.DocBin)
         assert len(train) == 100
 
+        train, test = rb_dataset.prepare_for_training(
```
Comment (frascuchon) [inline tests/client/test_dataset.py:644]: It would be nice to add also some tests to increase the overall coverage (train_size + test_size != 1, no train/test size with spark_nlp)

**Review by frascuchon — APPROVED** [review]
Nice work!

Just some insights  that are totally optional

File: src/argilla/client/datasets.py (hunk 1 above)
Comment (davidberenstein1957) [reply src/argilla/client/datasets.py:488]: nope, I export a pd.DataFrame, which can then be loaded into `spark-nlp` afterwards, because doing this directly requires a spark dataframe and session, which seemed like a bit too much overhead for me.

Comment (frascuchon) [reply src/argilla/client/datasets.py:488]: so, I think we should add this dep explicitly in our `pyproject.toml` since we're using it in the codebase

Comment (davidberenstein1957) [reply src/argilla/client/datasets.py:488]: Yes, but pandas is already in there and we don`t use  `spark-nlp`

Comment (frascuchon) [reply src/argilla/client/datasets.py:488]: ah ok it will be included by the panda dep. Good!

File: tests/client/test_dataset.py (hunk 6 above)
Comment (davidberenstein1957) [reply tests/client/test_dataset.py:644]: Thansk, I added this too.

**Review by frascuchon — APPROVED** [review]
